Emergency upgrades

afck opened the pull request:
They work just like regular upgrades except that we can't _download_ the immediate switch block: we have to create it ourselves, because after an emergency restart the pre-restart validators might not be trustworthy anymore.

Keeping #1665 open for now until some of the other fixes have been merged too, and can be tested together with this.

We also set the log level for `retrieved block header but block body is missing from database` to INFO, because fast sync is expected to download some headers without bodies.

Closes #2583.

afck committed:
Handle emergency restarts in fast sync.

They work just like regular upgrades except that we can't _download_ the
immediate switch block: we have to create it ourselves, because after an
emergency restart the pre-restart validators might not be trustworthy
anymore.

afck committed:
Update test: first full era is now era 1.

goral09 commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -725,13 +725,25 @@ pub(super) async fn run_chain_sync_task(
     let maybe_last_emergency_restart_era_id = chainspec.protocol_config.last_emergency_restart;
     if let Some(last_emergency_restart_era) = maybe_last_emergency_restart_era_id {
-        if last_emergency_restart_era > trusted_block_header.era_id() {
+        if last_emergency_restart_era > trusted_block_header.next_block_era_id() {
```
Maybe it's just me but `trusted_block_header.next_block_era_id() < ` would be easier 😅 

Also, can you add a comment that we have to increment the era id b/c there will be a no-block era with a switch block created just after the emeregency upgrade?

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2580/commits/544c44454fe753821eb4343d5a0d395286a02345.

afck committed:
Address review comments.

rafal-ch commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -725,13 +725,30 @@ pub(super) async fn run_chain_sync_task(
         }
+        // If the trusted hash specifies the last block before the emergency restart, we have to
+        // compute the immediate switch block ourselves, since there's no other way to verify that
+        // block. We just sync the trie there and and return, so the upgrade can be applied.
```
```suggestion
        // block. We just sync the trie there and return, so the upgrade can be applied.
```

afck replied:
Done in https://github.com/casper-network/casper-node/pull/2580/commits/6f42ce72d59aece5f145be3d7b92264f1eb7efcf.

afck committed:
Fix typo.

afck committed:
Lower missing block body warning level to INFO.

With fast sync it is expected that for some blocks we will only download
headers, not bodies.

afck committed:
Merge remote-tracking branch 'upstream/feat-fast-sync' into emergency_upgrades

rafal-ch reviewed (approved):
LGTM :+1:
